### PR TITLE
[mlir][scf] Add reductions support to `scf.parallel` fusion

### PR DESCRIPTION
**Pull request by Hardcode84**
Properly handle fusion of loops with reductions:
* Check there are no first loop results users between loops
* Create new loop op with merged reduction init values
* Update `scf.reduce` op to contain reductions from both loops
* Update loops users with new loop results

### CODE REVIEW

**Comment by Hardcode84** [comment]
@matthias-springer FYI, this will require some changes after reductions refactoring is merged

File: mlir/test/Dialect/SCF/parallel-loop-fusion.mlir [hunk 1]
```diff
@@ -387,3 +387,125 @@ func.func @do_not_fuse_alias(%A: memref<2x2xf32>, %B: memref<2x2xf32>,
...
+      %1 = arith.mulf %lhs, %rhs : f32
+      scf.reduce.return %1 : f32
+    }
+    scf.yield
+  }
+  return %res1, %res2 : f32, f32
+}
+
+// CHECK-LABEL: func @fuse_reductions
+//  CHECK-SAME:  (%[[A:.*]]: memref<2x2xf32>, %[[B:.*]]: memref<2x2xf32>)
```
Comment (etiotto) [inline mlir/test/Dialect/SCF/parallel-loop-fusion.mlir:421]: Add also the return type here?

**Comment by Hardcode84** [comment]
Updated to new reductions format

**Comment by Hardcode84** [comment]
ping

File: mlir/lib/Dialect/SCF/Transforms/ParallelLoopFusion.cpp [hunk 2]
```diff
@@ -131,29 +131,83 @@ static bool isFusionLegal(ParallelOp firstPloop, ParallelOp secondPloop,
...
+  firstToSecondPloopIndices.map(block1->getArguments(), block2->getArguments());
 
   if (!isFusionLegal(firstPloop, secondPloop, firstToSecondPloopIndices,
                      mayAlias))
     return;
 
-  b.setInsertionPointToStart(secondPloop.getBody());
-  for (auto &op : firstPloop.getBody()->without_terminator())
-    b.clone(op, firstToSecondPloopIndices);
+  DominanceInfo dom;
```
Comment (rengolin) [inline mlir/lib/Dialect/SCF/Transforms/ParallelLoopFusion.cpp:147]: The code could do with a bit more comments. The original code was clear enough by the function name, this new version isn't.

**Review by rengolin — COMMENTED** [review]
Please expand a bit the commit message and the comments throughout the code. It may be obvious to you or even someone who worked in this piece of code before, but it won't be to people who may have to bisect or revert this commit for some CI trouble.

File: mlir/lib/Dialect/SCF/Transforms/ParallelLoopFusion.cpp [hunk 3]
```diff
@@ -131,29 +131,83 @@ static bool isFusionLegal(ParallelOp firstPloop, ParallelOp secondPloop,
...
 }
 
 void mlir::scf::naivelyFuseParallelOps(
     Region &region, llvm::function_ref<bool(Value, Value)> mayAlias) {
   OpBuilder b(region);
   // Consider every single block and attempt to fuse adjacent loops.
+  SmallVector<SmallVector<ParallelOp>, 1> ploopChains;
   for (auto &block : region) {
-    SmallVector<SmallVector<ParallelOp, 8>, 1> ploopChains{{}};
+    ploopChains.clear();
```
Comment (rengolin) [inline mlir/lib/Dialect/SCF/Transforms/ParallelLoopFusion.cpp:208]: This change is unnecessary. 

The `SmallVector` empty constructor is basically a 2-word copy, with both `clear()` and `push_back()` already happening on c-tor/d-tor.

https://github.com/llvm/llvm-project/blob/aba40fb34a27f1d36b4b541bf04d414dca7d1c4c/llvm/include/llvm/ADT/SmallVector.h#L63

File: mlir/lib/Dialect/SCF/Transforms/ParallelLoopFusion.cpp [hunk 4]
```diff
@@ -131,29 +131,83 @@ static bool isFusionLegal(ParallelOp firstPloop, ParallelOp secondPloop,
...
+  b.setInsertionPoint(secondPloop);
+  auto newSecondPloop = b.create<ParallelOp>(
+      secondPloop.getLoc(), secondPloop.getLowerBound(),
+      secondPloop.getUpperBound(), secondPloop.getStep(), newInitVars);
+
+  Block *newBlock = newSecondPloop.getBody();
+  auto term1 = cast<ReduceOp>(block1->getTerminator());
+  auto term2 = cast<ReduceOp>(block2->getTerminator());
+
+  b.inlineBlockBefore(block2, newBlock, newBlock->begin(),
```
[the remark this answers is not in the document]
Comment (Hardcode84) [reply mlir/lib/Dialect/SCF/Transforms/ParallelLoopFusion.cpp:168]: We need to change results count, when merging `scf.parallel`s with reductions, the only way is to recreate op.

File: mlir/lib/Dialect/SCF/Transforms/ParallelLoopFusion.cpp (hunk 3 above)
Comment (Hardcode84) [reply mlir/lib/Dialect/SCF/Transforms/ParallelLoopFusion.cpp:208]: I know, the idea was to reuse same vector between loop iterations, avoiding possible allocations/deallocations if vector was big enough to be allocated in heap.

File: mlir/lib/Dialect/SCF/Transforms/ParallelLoopFusion.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (rengolin) [reply mlir/lib/Dialect/SCF/Transforms/ParallelLoopFusion.cpp:170]: Can you not just re-generate the `scf.reduction` inside the second loop?

[the remark this answers is not in the document]
Comment (Hardcode84) [reply mlir/lib/Dialect/SCF/Transforms/ParallelLoopFusion.cpp:170]: Each reduction corresponds to the parent `scf.parallel` op result value, so if the first loop had any reductions, those results must be part of the fused parent op, changing total results count.

File: mlir/test/Dialect/SCF/parallel-loop-fusion.mlir [hunk 5]
```diff
@@ -480,3 +480,182 @@ func.func @do_not_fuse_multiple_stores_on_diff_indices(
...
+    scf.reduce(%sum : f32) {
+    ^bb0(%lhs: f32, %rhs: f32):
+      %1 = arith.mulf %lhs, %rhs : f32
+      scf.reduce.return %1 : f32
+    }
+  }
+  return %res1, %res2 : f32, f32
+}
+
+// %res1 is used inside second scf.parallel, cannot fuse
```
Comment (etiotto) [inline mlir/test/Dialect/SCF/parallel-loop-fusion.mlir:658]: Add:
 - a test to prove that the 2 loops can be fused even if there are instructions between the loops that do not use the result produced by the first loop
 - a test to verify that the 2 loops aren't fused if an instruction in between the loops uses the first loop result

Comment (Hardcode84) [reply mlir/test/Dialect/SCF/parallel-loop-fusion.mlir:684]: done

**Review by etiotto — APPROVED** [review]

**Pull request merged**